activesupport/lib/active_support/message_encryptor.rb

```diff
@@ -208,6 +218,33 @@ def _decrypt(encrypted_message, purpose)
         raise InvalidMessage
       end
 
+      def iv_length_in_base64
+        @iv_length_in_base64 ||= ((4 * new_cipher.iv_len / 3) + 3) & ~3
+      end
+
+      def auth_tag_and_iv_indexes_for(encrypted_message)
+        if aead_mode?
+          auth_tag_index = encrypted_message.length - AUTH_TAG_LENGTH_IN_BASE64 - SEPARATOR_LENGTH
+          return if auth_tag_index < SEPARATOR_LENGTH || encrypted_message[auth_tag_index, SEPARATOR_LENGTH] != SEPARATOR
+        end
+
+        iv_index = (auth_tag_index || encrypted_message.length) - iv_length_in_base64 - SEPARATOR_LENGTH
+        return if iv_index.negative? || encrypted_message[iv_index, SEPARATOR_LENGTH] != SEPARATOR
+
+        [auth_tag_index, iv_index]
+      end
+
+      def get_encrypted_data_and_iv_and_auth_tag_from(encrypted_message)
+        auth_tag_index, iv_index = auth_tag_and_iv_indexes_for(encrypted_message)
+        return if iv_index.nil? || (aead_mode? && auth_tag_index.nil?)
+
+        encrypted_data = encrypted_message[0...iv_index]
```

Comment:
```suggestion
        encrypted_data = encrypted_message[0, iv_index]
```

This should be faster because it doesn't allocate a Range instance.

```
$ ruby -rbenchmark/ips -e "s = 'abracadabra'; n = 3; Benchmark.ips { |x| x.report('s[0...n]') { s[0...n] }; x.report('s[0, n]') { s[0, n] }; x.compare! }"
Warming up --------------------------------------
            s[0...n]   395.527k i/100ms
             s[0, n]   695.406k i/100ms
Calculating -------------------------------------
            s[0...n]      3.944M (± 1.1%) i/s -     19.776M in   5.014851s
             s[0, n]      6.919M (± 1.7%) i/s -     34.770M in   5.026923s

Comparison:
             s[0, n]:  6918908.1 i/s
            s[0...n]:  3944079.4 i/s - 1.75x  (± 0.00) slower
```

Reply:
Thanks! :raised_hands: